test_rmw_implementation/test/testing_macros.hpp

```diff
@@ -0,0 +1,31 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TESTING_MACROS_HPP_
+#define TESTING_MACROS_HPP_
+
+#include <chrono>
+#include <thread>
+
+/// Retry until `timeout` expires, sleeping for `delay` in between attempts.
+/*
+ * \note Time is measured against OS provided steady clock.
+ */
+#define SLEEP_AND_RETRY_UNTIL(delay, timeout) for ( \
+    auto loop_start_time = std::chrono::steady_clock::now(), \
+    iteration_start_time = loop_start_time; \
```

Comment:
Why is `iteration_start_time` needed? Why can't you just use `std::this_thread::sleep_for(delay)` instead of `sleep_until`?

Reply:
To avoid the time it takes to actually execute the loop code from adding up. Also, [`std::this_thread::sleep_for`](https://en.cppreference.com/w/cpp/thread/sleep_for) makes no guarantees as to which clock is used.

Reply:
I guess I mean something like:
```c++
#define SLEEP_AND_RETRY_UNTIL(delay, timeout) \
  for (auto start = std::chrono::steady_clock::now(); \ 
       std::chrono::steady_clock::now() - start < timeout; \
       std::this_thread::sleep_for(delay))
```

If you'd rather leave it as is, then I'm fine with it. It just seems like your use case doesn't require such precision.

Reply:
I didn't want to lose retries due to non-negligible execution times, but it is true that these timeouts are pretty arbitrary. 

Sure, why not. See https://github.com/ros2/rmw_implementation/pull/119/commits/4d85be9ec0d7244681f16beec420e5b2b0666127.